cpp/include/cudf/join.hpp

```diff
@@ -367,5 +367,138 @@ std::unique_ptr<cudf::table> cross_join(
   cudf::table_view const& right,
   rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
 
+/**
+ * @brief Hash join that builds hash table in creation and probes results in subsequent `*_join`
+ * member functions.
+ *
+ * This class enables the hash join scheme that builds hash table once, and probes as many times as
+ * needed (possibly in parallel).
+ */
+class hash_join {
+ public:
+  hash_join() = delete;
+  ~hash_join();
+  hash_join(hash_join const&) = delete;
+  hash_join(hash_join&&)      = delete;
+  hash_join& operator=(hash_join const&) = delete;
+  hash_join& operator=(hash_join&&) = delete;
+
+  /**
+   * @brief Construct a hash join object and build the internal hash table used for subsequent
+   * probe calls.
+   *
+   * @param build The build table, from which the hash table is built.
+   * @param build_on The column indices from `build` to join on.
+   */
+  hash_join(cudf::table_view const& build, std::vector<size_type> const& build_on);
+
+  /**
+   * @brief Side of the probe table in the joined table. Only applicable for inner join.
+   */
+  enum class probe_output_side { LEFT, RIGHT };
+
+  /**
+   * @brief Performs an inner join by probing in the internal hash table.
+   *
+   * Given that it is sometimes desired to choose the small table to be the `build` side for an
+   * inner join (which is automatically done in cudf::innner_join()), it is handy to be able to
+   * specify which side in the joined table the `probe` portion is on. That is, one can check
+   * whether to use `left` or `right` table to be the build side by comparing their row counts ahead
+   * of `hash_join` object, but still preserve the logical order of the `left`/`right` in the joined
+   * table. For instance, if the `left` table is the small one (and thus the `build` side),
+   * specifying the `probe_output_side` to `RIGHT` will output the `right` table (probe side) on the
+   * right side in the joined table.
```

Comment:
So this parameter is kind of awkward. I think we could eliminate it by making these member functions return a `pair<unique_ptr<table>, unique_ptr<table>>` of the columns from the `build` and `probe` tables respectively. Then at the call site they can be swapped if necessary.

Reply:
I think caller can always adjust the order of `build`/`probe` columns, and it is not much harder than adjusting the order of a packed `build`/`probe` table pair.

If the concern is to keep the API simple, maybe we can just remove this parameter?

Reply:
Yeah, I think the `probe_output_side` parameter is confusing. 

So you're suggesting return a single `table` and just switch the order of the columns in that table? I think that can be tricky, especially due to the `columns_in_column` stuff.

Reply:
I mean returning a single table and the columns order being **strictly** of `probe` on the left and `build` on the right. Users can adjust columns order within the returned table.

Reply:
Additionally, I think returning `build`/`probe` table pair would be overkill, since getting a single result table from a `*_join` call seems like a more common case.

Reply:
> Users can afterwards adjust columns order within the returned table on their own demands.

But I think `columns_in_common` make it tricky to know what columns correspond to `build` vs `probe`. Maybe I'm mistaken. Could you try implementing the existing `cudf::inner_join` w/o the `probe_output_side` parameter and manually rearranging the columns to show what it would take?

Reply:
I think `columns_in_common` can still go as described in the old `cudf::inner/left/full_join` functions, i.e. `build` always being right and `probe` being left and `columns_in_common` being gathered from `probe` (left).

The rearranging code would be like:
`
table_view t0(a0, b0, c0);
hash_join hj(t0, {0, 1});
table_view t1(a1, b1, c1);
result = hj.inner_join(t1, {0, 1}, {{0, 0}, {1, 1}});
// Refer to the description of `columns_in_common`, user will know result is (a1, b1, c1, c0).
// Switching left/right into a new table view will be like:
table_view rearranged(result[0], result[1], result[3], result[2]);
`
Does this answer the question?

Reply:
We need to implement that logic when implementing the existing `cudf::inner_join` in terms of the new `hash_join` OO implementaiton.

Reply:
> We need to implement that logic when implementing the existing `cudf::inner_join` in terms of the new `hash_join` OO implementation.

Sorry, I don't think I'm following. Could you explain more?

Reply:
I think I got ya. Let me think a moment.

Reply:
You were right. It will be non-trivial to implement the existing `cudf::inner_join` using `hash_join` and rearranging the resulting table.

But I'm thinking returning `build`/`probe` table pair is equally non-trivial in terms of rearranging, since the "common" columns have to be in either side (but not both, otherwise they will be materialized twice), the trouble of `columns_in_common` still exists in the same way as in the single resulting table.

Moreover, returning `build`/`probe` table pair makes `hash_join::inner_join` not quite aligned with the other two `hash_join::left_join` and `hash_join::full_join` members in terms of signature, and users pay more attention of combining them despite they want to rearrange them or not.

What do you think?

Reply:
> since the "common" columns have to be in either side (but not both, otherwise they will be materialized twice), the trouble of columns_in_common still exists in the same way as in the single resulting table.

Won't this be taken care of automatically be swapping the values in `columns_in_common`? 

> Moreover, returning build/probe table pair makes hash_join::inner_join not quite aligned with the other two hash_join::left_join and hash_join::full_join members in terms of signature,

I'm not too concerned about this. This `hash_join` object is already intended as more of an "expert-mode" API anyways.

Reply:
Oh, let me explain more about the "non-trivial" part.

Image that we end up with the following `hash_join::inner_join` API (w/o the `probe_output_side` parameter):

`std::pair<table /*probe portion*/, table /*build portion*/> hash_join::inner_join(probe, probe_on, columns_in_common, mr);`

1. The order of probe/build portions needs to be fixed as there is no flag in the parameter list to specify it. We can inform users this spec through proper comments. This wouldn't be an issue.
2. Due to the existence of `columns_in_common`, the "common" columns need to go into either probe or build portion. We can too fix this spec through comments and inform users, say we go with "common columns will be output in the probe portion".

Now consider the old `cudf::inner_join` API, under the circumstances that `left` being smaller:
```
hash_join(left, left_on);
swap(columns_in_common);
[probe, build] = hash_join.inner_join(right, right_on, columns_in_common);
// Now do rearrangement to preserve the logical left/right order.
probe_common_indices = ...;
probe_noncommon_indices = ...;
common_columns = probe.select(probe_common_indices);
right_noncommon_columns = probe.select(probe_noncommon_indices);
left_noncommon_columns = build;
left_columns = ...; // Zip `left_noncommon_columns` with `common_columns` to restore the relative orders of the columns in the original `left` table.
return table(combine(left_columns, right_noncommon_columns));
```

Likewise, for the direct `hash_join::inner_join` users, once they want to explicitly swap the build/probe tables, they'll have to do the last 8 lines of the above code block. IMO this would be cumbersome even for "expert" users.

Please help to see if my explanation make sense.

Reply:
Any comment?

Reply:
@jrhemstad Hi Jake, could you please take a look the comments above? Thanks.

Reply:
I had to walk through a concrete example before I could see what the issue is. Here's the example I went through:

Imagine two tables:
```
table p : columns {a, b}
table q : columns {d, a}
```

We want to join "on" the `a` column from each table. Furthermore, we want to follow Pandas' semantics and only materialize one copy of `a` since it has the same name. 

So calling `cudf::inner_join` would look like:
```
cudf::inner_join(p, q, left_on{0}, right_on{1}, common{ {0,1} })
```

Since `p` is the "left" table, we want to keep the copy of `a` with `p`. 

Now we have two situations to consider.

First, consider `p.num_rows() > q.num_rows()`. 

`q` is the smaller table, so we build on `q`. This is swapping the original (p,q) order, so we need to take care of that swapping.

```
[q', p'] = hash_join obj(q, on{1}).inner_join(p, on{0}, common{1,0});
```

The tables `q'` and `p'` hold the join results from `q` and `p` respectively. 

**So here is the problem:** If we say "common columns always go with the build table", then `q'` will contain the `a` column, but we wanted it in `p'`! Instead, if we say "common columns always go with the probe table", then it will break in the situation where `p.num_rows() < q.num_rows()`. 

So sadly we do need some kind of parameter that controls where the "common" columns go, but this has become very convoluted. 

My preference would still to make `inner_join` return a `pair<table,table>` for build,probe respectively, but change the `probe_output_side` argument to instead control whether common columns are placed in `build` or `probe`.

Reply:
I'll open an issue to get rid of `columns_in_common` all together by instead having APIs to simply return the indices into the left and right table of the matching rows.

Reply:
So at least we both now agree that an extra parameter would be necessary to ease the rearrangement of common columns in the output table(s): either `probe_output_side` or something like `common_columns_go_with` as you proposed.

I can do the `common_columns_go_with` way. But I'm not sure about your latest comment about getting rid of `columns_in_common`. I think changing this PR to the `common_columns_go_with` way could be done in parallel with getting rid of `columns_in_common`. If so, I'll go on changing the code. Or should I wait for your new APIs and leverage them after they are done?

Please advice. Thanks.

Reply:
> I think changing this PR to the common_columns_go_with way could be done in parallel with getting rid of columns_in_common. If so, I'll go on changing the code. Or should I wait for your new APIs and leverage them after they are done?

Getting rid of `columns_in_common` will be future work and should not block this PR. Proceed with making the minimal changes to this PR and we'll get it merged.

Reply:
Addressed.